Dinah.Core/_IO/JsonFilePersister.cs

```diff
@@ -119,46 +115,67 @@ private void saveFile(object? _, EventArgs __)
 				return;
 			}
 
-			lock (_locker)
-            {
-                try
-                {
-					OnSaving();
-
-					if (JsonPath is null)
-					{
-						File.WriteAllText(Path, JsonConvert.SerializeObject(Target, Formatting.Indented, GetSerializerSettings()));
-						return;
-					}
-
-					// path must
-					// - exist
-					// - have valid jsonPath match
-
-					var contents = File.ReadAllText(Path);
-					var allToken = JObject.Parse(contents);
-					var pathToken = allToken.SelectToken(JsonPath);
-
-					if (pathToken is null)
-						throw new JsonSerializationException($"No match found at JSONPath: {JsonPath}");
-
-					// load existing identity into JObject
-					var serializer = JsonSerializer.Create(GetSerializerSettings());
-					var idJObj = JObject.FromObject(Target, serializer);
-
-					// replace. this propgates to 'allToken'
-					pathToken.Replace(idJObj);
-
-					var allSer = JsonConvert.SerializeObject(allToken, Formatting.Indented, GetSerializerSettings());
-					File.WriteAllText(Path, allSer);
-				}
-                finally
-                {
-					OnSaved();
-				}
+			try
+			{
+				OnSaving();
+				var locker = _locks[Path];
+                lock (locker)
+                    saveJson();
+            }
+			finally
+			{
+				OnSaved();
 			}
 		}
 
+		private void saveJson()
+		{
+			if (JsonPath is null)
+			{
+				serializeAndWrite(Target);
+				return;
+			}
+
+			// path must
+			// - exist
+			// - have valid jsonPath match
+
+			var contents = File.ReadAllText(Path);
+			var allToken = JObject.Parse(contents);
+			var pathToken = allToken.SelectToken(JsonPath) ?? throw new JsonSerializationException($"No match found at JSONPath: {JsonPath}");
+
+			// load existing identity into JObject
+			var serializer = JsonSerializer.Create(GetSerializerSettings());
+			var idJObj = JObject.FromObject(Target, serializer);
+
+			// replace. this propgates to 'allToken'
+			pathToken.Replace(idJObj);
+
+			serializeAndWrite(allToken);
+		}
+
+		private void serializeAndWrite(object payload)
+		{
+			// prevent invalid writes
+			var json = JsonConvert.SerializeObject(payload, Formatting.Indented, GetSerializerSettings());
+			if (string.IsNullOrWhiteSpace(json))
+				throw new JsonSerializationException("Could not write json file. Empty payload");
+
+			preventRapidWrites();
+
+			File.WriteAllText(Path, json);
+		}
+
+		private static DateTime lastWrite = DateTime.MinValue;
+		// prevent multiple writes in quick succession
+		private void preventRapidWrites()
+		{
+			if (DateTime.UtcNow - lastWrite < TimeSpan.FromMilliseconds(100))
+				Thread.Sleep(100);
+
+			lastWrite = DateTime.UtcNow;
+		}
```

Comment:
This will prevent 2 quick succession writes, but not more than 2. If N writes are started in quick succession, then...
- Write 1 goes through immediately and sets `lastWrite`
- Write 2 sees the recent `lastWrite` and triggers Thread.Sleep(100);
- Write 3 sees the recent `lastWrite` and triggers its own Thread.Sleep(100);
- The time delta between Write 2's sleep and write 3's sleep is equal to the time delta between their initial write requests, so they continue and there is no net delay between them

Reply:
Hmm, if they're on the same object, the lock should take care of running them in order and hitting that part in order. If not on the same object, then they'll have diff lastWrite state

Reply:
New commit makes `lastWrite` static per path

Reply:
> If not on the same object, then they'll have diff lastWrite state

lastWrite is static.

But you are right. If they're on the same object, file locks take care of it. If they're not on the same object but they do have the same file name, then the dictionary locks take care of it.

The only way multiple successive writes can happen is if there are different persister instances with different filenames, but we probably don't care about that, right?

Reply:
> New commit makes `lastWrite` static per path

I should have refreshed before submitting my last comment.

Reply:
> The only way multiple successive writes can happen is if there are different persister instances with different filenames, but we probably don't care about that, right?

Assuming different absolute paths (checking absolute path that felt overkill for this purpose), then what you described is different files getting updated at different times. This should always be allowed. Anything else would be a bug. And looking over the history of my changes today: yeah, static `lastWrite` was a bug. There's no reason to throttle file A just because file B was recently written. This is horrible. See previous comment about lack of sleep. Thanks for your review.